File: ch4/clip5/boj2294/main.cpp
#include <iostream>

int main()
{
    int N, K;
    int coin;
    int numberOfType[10001];
    
    scanf("%d %d", &N, &K);

    numberOfType[0] = 0;
    for(int i = 1; i <= K; i++)
    {
        numberOfType[i] = 10001;
    }

    for(int i = 0; i < N; i++)
    {
        scanf("%d", &coin);

        for(int k = coin; k <= K; k++)
        {
            numberOfType[k] = std::min(numberOfType[k], numberOfType[k - coin] + 1);
        }
    }

    if(10001 == numberOfType[K])
    {
        printf("-1");
    }
    else
    {
        printf("%d", numberOfType[K]);
    }

    return 0;
}

File: ch4/clip5/boj2294/main.py
n, k = map(int, input().split())
coins = [int(input()) for _ in range(n)]
dp = [100001] * (k + 1)

for i in range(n):
    coin = coins[i]
    if coin <= k:
        dp[coin] = 1

for i in range(1, k + 1):
    for j in range(n):
        if i + coins[j] <= k:
            dp[i + coins[j]] = min(dp[i + coins[j]], dp[i] + 1)
        if i - coins[j] >= 0:
            dp[i] = min(dp[i], dp[i - coins[j]] + 1)

if dp[k] == 100001:
    print(-1)
else:
    print(dp[k])
